nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/StreamBulkScroller.java

```diff
@@ -0,0 +1,173 @@
+        String kvStore = service.getKeyValueStore();
+        List<String> operations = service.getOperations();
+        List<String> mapping = new ArrayList<>();
+        mapping.add("i1:" + SET_STREAM_NAME);
+        int i = 1;
+        for (String operation : operations) {
+            mapping.add(String.format("o%s:%s", i, operation));
+            i++;
+        }
+        return Topology.builder()
+                       .addComputation( //
+                               () -> new BulkDocumentScrollerComputation(COMPUTATION_NAME, operations.size(), kvStore,
+                                       scrollBatchSize, scrollKeepAliveSeconds), //
+                               mapping)
+                       .build();
+    }
+
+    public static class BulkDocumentScrollerComputation extends AbstractComputation {
+
+        protected final KeyValueStore kvStore;
+
+        protected final int scrollBatchSize;
+
+        protected final int scrollKeepAliveSeconds;
+
+        public BulkDocumentScrollerComputation(String name, int nbOutputStreams, String kvStore, int scrollBatchSize,
+                int scrollKeepAliveSeconds) {
+            super(name, 1, nbOutputStreams);
+            this.kvStore = Framework.getService(KeyValueService.class).getKeyValueStore(kvStore);
+            this.scrollBatchSize = scrollBatchSize;
+            this.scrollKeepAliveSeconds = scrollKeepAliveSeconds;
+        }
+
+        @Override
+        public void processRecord(ComputationContext context, String inputStreamName, Record record) {
+            TransactionHelper.runInTransaction(() -> processRecord(context, record));
+        }
+
+        protected void processRecord(ComputationContext context, Record record) {
+            try {
+                BulkCommand command = getBulkCommandJson(record.data);
+                if (!kvStore.compareAndSet(record.key + STATE, SCHEDULED.toString(), BUILDING.toString())) {
+                    log.error("Discard record: " + record + " because it's already building");
+                    return;
+                }
+                try (CloseableCoreSession session = CoreInstance.openCoreSession(command.getRepository(),
+                        command.getUsername())) {
+                    // scroll documents
+                    ScrollResult<String> scroll = session.scroll(command.getQuery(), scrollBatchSize,
+                            scrollKeepAliveSeconds);
+                    long documentCount = 0;
+                    while (scroll.hasResults()) {
+                        List<String> docIds = scroll.getResults();
+                        // send these ids as keys to the appropriate stream
+                        // key will be bulkOperationId/docId
+                        // value/data is a BulkCommand serialized as JSON
+                        docIds.forEach(docId -> context.produceRecord(command.getOperation(), record.key + '/' + docId,
+                                record.data));
```

Comment:
ok for the first impl but adding the command in each record is too much overhead when having millions of ids. An option for next version is to output record with multiple ids. At this point I don't know if we need a key instead of simply roundrobin on partition and if we need the command.

Reply:
Indeed, we'll check if we really need command in the bulk operation.
I'm not agree to send multiple ids in one record. For case like conversion, some can take longer than others. A list of ids could break a bit sharding.